### PR TITLE
feat: `VariableSizeSubspace`

### PR DESCRIPTION
**Pull request by andiwand**
As an alternative to the existing `Acts::detail::FixedSizeSubspace` I implement `VariableSizeSubspace` which is very similar except that the actual size of the subspace can be runtime dependent.

Pulled out of https://github.com/acts-project/acts/pull/3107

### CODE REVIEW

File: Core/include/Acts/Utilities/detail/Subspace.hpp [hunk 1]
```diff
@@ -204,6 +210,175 @@ class FixedSizeSubspace {
...
+  // n<=8, this still takes only 64bit of memory.
+  std::array<std::uint8_t, kFullSize> m_axes{};
+
+ public:
+  /// Construct from a container of axis indices.
+  ///
+  /// @tparam index_t Input index type, must be convertible to std::uint8_t
+  /// @param indices Unique, ordered indices
+  template <typename index_t, std::size_t kSize>
+  constexpr VariableSizeSubspace(const std::array<index_t, kSize>& indices) {
```
Comment (paulgessinger) [inline Core/include/Acts/Utilities/detail/Subspace.hpp:249]:
```suggestion
  explicit constexpr VariableSizeSubspace(const std::array<index_t, kSize>& indices) {
```

File: Core/include/Acts/Utilities/detail/Subspace.hpp [hunk 2]
```diff
@@ -204,6 +210,175 @@ class FixedSizeSubspace {
...
+  std::uint64_t fullProjectorBits() const {
+    std::uint64_t result = 0;
+
+    for (std::size_t i = 0; i < kFullSize; ++i) {
+      for (std::size_t j = 0; j < kFullSize; ++j) {
+        // the bit order is defined in `Acts/Utilities/AlgebraHelpers.hpp`
+        // in `matrixToBitset`
+        std::size_t index = kFullSize * kFullSize - 1 - (i + j * kFullSize);
+        if (i < m_size && m_axes[i] == j) {
+          result |= (1ull << index);
```
Comment (paulgessinger) [inline Core/include/Acts/Utilities/detail/Subspace.hpp:373]: Same here.

File: Core/include/Acts/Utilities/detail/Subspace.hpp [hunk 3]
```diff
@@ -204,6 +210,175 @@ class FixedSizeSubspace {
...
+/// @tparam kSize Size of the subspace
+template <std::size_t kFullSize>
+class VariableSizeSubspace {
+  static_assert(kFullSize <= static_cast<std::size_t>(UINT8_MAX),
+                "Full vector space size is larger than the supported range");
+
+  template <typename scalar_t>
+  using ProjectionMatrix = Eigen::Matrix<scalar_t, Eigen::Dynamic, kFullSize>;
+  template <typename scalar_t>
+  using ExpansionMatrix = Eigen::Matrix<scalar_t, kFullSize, Eigen::Dynamic>;
```
Comment (paulgessinger) [inline Core/include/Acts/Utilities/detail/Subspace.hpp:225]: My one concern here is that we should avoid making the dynamic matrices "viral", so they don't spread to other parts of the code.

I wonder if we should remove or discourage using the projection and expansion matrices completely, and use methods that return statically sized objects instead. What do you think?

Comment (andiwand) [reply Core/include/Acts/Utilities/detail/Subspace.hpp:225]: That's a good point. Removing it or only passing the matrix in as template param might be better I guess? Like `template<MatrixType> fillProjector(MatrixType &matrix)` and then asserting on the dimensions.

File: Core/include/Acts/Utilities/detail/Subspace.hpp (hunk 2 above)
Comment (andiwand) [reply Core/include/Acts/Utilities/detail/Subspace.hpp:373]: Here it is an actual bitwise operation. Does it also complain about that?

File: Core/include/Acts/Utilities/detail/Subspace.hpp (hunk 3 above)
Comment (paulgessinger) [reply Core/include/Acts/Utilities/detail/Subspace.hpp:220]: That would work. Alternatively, you could require supplying an explicit dimension template parameter, although that makes it a bit more annoying.

Comment (andiwand) [reply Core/include/Acts/Utilities/detail/Subspace.hpp]: I removed it since it is not used anyways for now

**Review by paulgessinger — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :red_circle: Athena integration test results [10344f536a93ed22997e6dd8ffec9911477a250d]
### :red_circle: Some tests have failed!
Please investigate the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/7656796)!

|status|job|report|
|------|---|------|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608465)||
|:green_circle: | [test_ActsDumpGeometryIdentifiers](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608464)||
|:green_circle: | [test_ActsCheckObjectCountsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608463)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608462)||
|:green_circle: | [test_ActsPersistifySeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608461)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608460)||
|:green_circle: | [test_ActsAnalogueClustering](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608459)||
|:green_circle: | [test_ActsWorkflowHeavyIons](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608458)||
|:green_circle: | [test_ActsWorkflowFastTracking](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608457)||
|:green_circle: | [test_ActsWorkflowCached](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608456)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608455)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608454)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608453)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608452)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608451)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608450)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608449)||
|:green_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608448)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608447)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608446)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608445)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608444)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608443)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608442)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608441)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608440)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608439)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608438)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608437)||
|:red_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608436)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40608435) | [:paperclip:](https://acts-athena-outputs.web.cern.ch/test_ttbarPU40_reco/pulls/3331/)|

**Comment by paulgessinger** [comment]
Athena test failure could be an unrelated filesystem issue. I triggered a rerun.

**Review by paulgessinger — APPROVED** [review]

**Pull request merged**